docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,275 @@
+    trusted {
+        public void ecall_with_user_check(
+            [user_check] void* blob);
+    };
+};
+```
+In this simplified EDL example, the function takes one pointer-to-void parameter. Remember that the primary benefit of using supported types in the EDL is that the generated boiler-plate code performs the necessary and secure marshaling for the developer. When developers specify that a parameter is `user_check`, they are signaling to oeedger8r that their application code will perform the marshaling. (Two examples of when custom marshaling is helpful are 1) when data formats are dynamic, and 2) when sharing large blocks of memory -- custom marshaling avoids the boiler-plate code's intermediate copying of data into safe in-enclave buffers.) 
+
+> We'd like to stress that a secure best practice for enclave interfaces is to _avoid_ custom marshaling. But we understand that there may be cases where avoiding it is impractical, so in these sections we will help developers do it safely.
+
+The ecall function implementation below has several problems. We will be correcting this code as we move through the material. The general impetus for custom marshaling in this sample code is to avoid expensive intermediate copying of shared memory introduced by the oeedger8r's generated code. (Data passed to `render_data()` is "flat", that is, the data format is non-referential and thus non-_self_-referencing. These attributes reduce threats to enclaves from untrusted data streams. More on this in the following sections.)
+
+```c++
+typedef struct _blob {
+    void* data;
+    size_t size;
+} blob_t;
+
+// WARNING: Portions of this code are intentionally flawed to demonstrate common pitfalls.
+int ecall_with_user_check0(void* ptr) {
+
+    blob_t* blob;
+
+    if (ptr == nullptr)
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // blobs to be rendered are limited to 4K
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+
+## 1.1 Ensuring memory is where it should be
+
+Remember that the code we are focused on is running _inside_ of the enclave. The host caller invoked this function from _outside_ the enclave, passing in a pointer to host memory containing well-formed data (in the non-malicious case). But in the Open Enclave security model, the host caller is untrusted - we must treat untrusted input with caution. 
+
+One important security check is to ensure that memory blocks are on the correct side of the security boundary. It's common for host applications to pass host memory blocks to the enclave, as is the case with this example code. The Open Enclave SDK provides two library functions for checking memory buffers: `oe_is_within_enclave()` and `oe_is_outside_enclave()`.
+
+Let's update the code to ensure that the memory block is at least the size of a `blob_t` and is located _strictly_ outside the secured enclave memory region before operating on it by calling `oe_is_outside_enclave()`.
```

Comment:
This section is not labeled as SGX specific, but the discussion appears to make an SGX-specific assumption that host memory can be directly read by the enclave.  This is not true in general, where line 94 below is unsafe because the memory pointed to by blob might not be readable from the encave.   If this intends to talk specifically about SGX again, then it should say so, like section 1.2 does.

Reply:
It's better to generalize this section, so I removed host-specific references.

Reply:
I'm beginning to waffle on this. Perhaps we should make the entire Custom Marshalling section explicitly SGX focused?

Reply:
After some mulling, the proper way to take this guidance is to make it general and hw-agnostic, following the spirit of Open Enclave.  The challenge though, is the potential mixed semantics of the memory models: This is evident in the different memory models of SGX and OP-TEE. When both are considered, it's difficult to choose the proper abstraction: SGX is satisfied by "outside" or "within" the enclave, but with OP-TEE it's "secure" or "nonsecure" -- the memory buffer could be shared, and thus neither necessarily "outside" nor "within" the enclave. 

BTW, I've always been bothered by the OE OP-TEE implementation of `oe_is_outside_enclave()`, and now I'm convinced it's incorrect in ways similar to why the SGX implementations are not boolean opposites. @dthaler, what do you think?
https://github.com/openenclave/openenclave/blob/cd72fd7069488ba6f453c8f5f47bd9fd9a6e6c0d/enclave/core/optee/bounds.c#L33-L36

Let me attempt to generalize the text in light of the comments above.

Reply:
Most of the points in the doc apply equally to SGX, OP-TEE, and any other TEEs, so it's good to keep the doc general where possible and just use TEE-specific examples.  For OP_TEE in particular, the discussion in #2139 is relevant to this doc... is_within_enclave() still makes sense as being secure, non-shared.  is_outside_enclave() is more nuanced as discussed in that issue.

Reply:
My personal view is that `oe_is_outside_enclave()` is a confusingly named API to begin with (for both SGX and OP-TEE) for the reasons articulated in the PR and in #2139.

Reply:
71be4bcb generalizes the text covering memory models.  I'd appreciate your continued review @dthaler.